pkg/il/builder.go

```diff
@@ -0,0 +1,278 @@
+	f.op1(ResolveD, f.id(n))
+}
+
+// TResolveDouble appends the "tresolve_d" instruction to the byte code.
+func (f *Builder) TResolveDouble(n string) {
+	f.op1(TResolveD, f.id(n))
+}
+
+// ResolveRecord appends the "resolve_r" instruction to the byte code.
+func (f *Builder) ResolveRecord(n string) {
+	f.op1(ResolveR, f.id(n))
+}
+
+// TResolveRecord appends the "tresolve_r" instruction to the byte code.
+func (f *Builder) TResolveRecord(n string) {
+	f.op1(TResolveR, f.id(n))
+}
+
+// IPushBool appends the "ipush_b" instruction to the byte code.
+func (f *Builder) IPushBool(b bool) {
+	f.op1(IPushB, boolToOpcode(b))
+}
+
+// IPushStr appends the "ipush_s" instruction to the byte code.
+func (f *Builder) IPushStr(s string) {
+	f.op1(IPushS, f.id(s))
+}
+
+// IPushInt appends the "ipush_i" instruction to the byte code.
+func (f *Builder) IPushInt(i int64) {
+	a1, a2 := integerToOpcode(i)
+	f.op2(IPushI, a1, a2)
+}
+
+// IPushDouble appends the "ipush_d" instruction to the byte code.
+func (f *Builder) IPushDouble(n float64) {
+	a1, a2 := doubleToOpcode(n)
+	f.op2(IPushD, a1, a2)
+}
+
+func boolToOpcode(b bool) uint32 {
+	i := uint32(0)
+	if b {
+		i = 1
+	}
+	return i
+}
+
+// Xor appends the "xor" instruction to the byte code.
+func (f *Builder) Xor() {
+	f.op0(Xor)
+}
+
+// EQString appends the "eq_s" instruction to the byte code.
+func (f *Builder) EQString() {
+	f.op0(EqS)
+}
+
+// IEQString appends the "ieq_s" instruction to the byte code.
+func (f *Builder) IEQString(v string) {
+	f.op1(IEqS, f.id(v))
+}
+
+// EQBool appends the "eq_b" instruction to the byte code.
+func (f *Builder) EQBool() {
+	f.op0(EqB)
+}
+
+// IEQBool appends the "ieq_b" instruction to the byte code.
+func (f *Builder) IEQBool(v bool) {
+	f.op1(IEqB, boolToOpcode(v))
+}
+
+// EQInteger appends the "eq_i" instruction to the byte code.
+func (f *Builder) EQInteger() {
+	f.op0(EqI)
+}
+
+// IEQInteger appends the "eq_i" instruction to the byte code.
+func (f *Builder) IEQInteger(v int64) {
+	a1, a2 := integerToOpcode(v)
+	f.op2(IEqI, a1, a2)
+}
+
+// EQDouble appends the "eq_d" instruction to the byte code.
+func (f *Builder) EQDouble() {
+	f.op0(EqD)
+}
+
+// IEQDouble appends the "ieq_d" instruction to the byte code.
+func (f *Builder) IEQDouble(v float64) {
+	a1, a2 := doubleToOpcode(v)
+	f.op2(IEqD, a1, a2)
+}
+
+// Not appends the "not" instruction to the byte code.
+func (f *Builder) Not() {
+	f.op0(Not)
+}
+
+// Or appends the "or" instruction to the byte code.
+func (f *Builder) Or() {
+	f.op0(Or)
+}
+
+// And appends the "and" instruction to the byte code.
+func (f *Builder) And() {
+	f.op0(And)
+}
+
+// Lookup appends the "lookup" instruction to the byte code.
+func (f *Builder) Lookup() {
+	f.op0(Lookup)
+}
+
+// ILookup appends the "ilookup" instruction to the byte code.
+func (f *Builder) ILookup(v string) {
+	f.op1(ILookup, f.id(v))
+}
+
+// AllocateLabel allocates a new label value for use within the code.
+func (f *Builder) AllocateLabel() string {
```

Comment:
You may need to reference a label before you need to declare it. i.e.:
jmp Foo
...
Foo:

In that case, you need to allocate the label first, use it in jmp and then set its position.